cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go

```diff
@@ -386,3 +358,141 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
 	}
 	return nil
 }
+
+// KubeConfigs is a mapping between kubeconfig file name and spec
+type KubeConfigs map[string]*kubeConfigSpec
+
+type kubeConfigsVisitor func(string, *kubeConfigSpec) error
+
+// executes the supplied visitor function with each key, value pair.
+func (o KubeConfigs) visit(visitor kubeConfigsVisitor) error {
+	if visitor == nil {
+		return fmt.Errorf("%w: visitor was nil", errInvalid)
+	}
+	for name, spec := range o {
+		if err := visitor(name, spec); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetDefaultKubeConfigs returns all KubeConfigSpecs without CA key or certificate.
+func GetDefaultKubeConfigs(cfg *kubeadmapi.InitConfiguration) (KubeConfigs, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("%w: cfg was nil", errInvalid)
+	}
+	return getKubeConfigSpecsBase(cfg)
+}
+
+func getKubeConfigSpecsBase(cfg *kubeadmapi.InitConfiguration) (KubeConfigs, error) {
+	apiServer, err := kubeadmutil.GetControlPlaneEndpoint(cfg.ControlPlaneEndpoint, &cfg.LocalAPIEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	return KubeConfigs{
+		kubeadmconstants.AdminKubeConfigFileName: {
+			APIServer:  apiServer,
+			ClientName: "kubernetes-admin",
+			ClientCertAuth: &clientCertAuth{
+				Organizations: []string{kubeadmconstants.SystemPrivilegedGroup},
+			},
+		},
+		kubeadmconstants.KubeletKubeConfigFileName: {
+			APIServer:  apiServer,
+			ClientName: fmt.Sprintf("%s%s", kubeadmconstants.NodesUserPrefix, cfg.NodeRegistration.Name),
+			ClientCertAuth: &clientCertAuth{
+				Organizations: []string{kubeadmconstants.NodesGroup},
+			},
+		},
+		kubeadmconstants.ControllerManagerKubeConfigFileName: {
+			APIServer:      apiServer,
+			ClientName:     kubeadmconstants.ControllerManagerUser,
+			ClientCertAuth: &clientCertAuth{},
+		},
+		kubeadmconstants.SchedulerKubeConfigFileName: {
+			APIServer:      apiServer,
+			ClientName:     kubeadmconstants.SchedulerUser,
+			ClientCertAuth: &clientCertAuth{},
+		},
+	}, nil
+}
+
+// kubeConfigAndCSRCreator is used in ExternalCA mode to create kubeconfig files
+// and adjacent CSR files.
+type kubeConfigAndCSRCreator struct {
+	kubeConfigDir string
+	kubeadmConfig *kubeadmapi.InitConfiguration
+}
+
+// create creates the the files.
+// Implements kubeConfigsVisitor.
+func (o *kubeConfigAndCSRCreator) create(name string, spec *kubeConfigSpec) error {
+	if o == nil {
+		return fmt.Errorf("%w: create was called on a nil pointer", errInvalid)
+	}
+	if name == "" {
+		return fmt.Errorf("%w: name was empty", errInvalid)
+	}
+	if spec == nil {
+		return fmt.Errorf("%w: spec was nil", errInvalid)
+	}
+	kubeConfigPath := filepath.Join(o.kubeConfigDir, name)
+	if _, err := os.Stat(kubeConfigPath); err == nil {
+		return fmt.Errorf("%w: kube config: %s", os.ErrExist, kubeConfigPath)
```

Comment:
`os.ErrExist` seems like a natural error to return here, but I considered adding a package level `errExist` type instead.